Add delta lake integration docs (#2052)
treeverse/lakeFS: docs/integrations/delta.md

---
layout: default
title: Delta Lake
description: This section explains how to use Delta Lake with lakeFS
parent: Integrations
nav_order: 62
has_children: false
---

# Using lakeFS with Delta Lake
{: .no_toc }

[Delta Lake](https://delta.io/) is an open file format designed to improve performance and provide transactional guarantees to data lake tables.

lakeFS is format-agnostic, so you can save data in Delta format within a lakeFS repository to get the benefits of both technologies. Specifically:

1. ACID operations can now span across many Delta tables.
1. [CI/CD hooks](../setup/hooks.md) can validate Delta table contents, schema, or even referential integrity.

> 1. lakeFS branches can zero-copy Delta tables for quick experimentation.

Reviewer: ```suggestion
1. lakeFS supports zero-copy branching for quick experimentation with full isolation.
```

## Table of contents
{: .no_toc .text-delta }

1. TOC
{:toc}

## Configuration

> Most commonly Delta tables are interacted with in a Spark environment given the native integration between Delta and Spark.

Reviewer: ```suggestion
Most commonly Delta tables are interacted with in a Spark environment given the native integration between Delta Lake and Spark.
```

To configure a Spark environment to read from and write to a Delta table within a lakeFS repository, we need to set the proper credentials and endpoint in the S3 Hadoop configuration, like we do with any [Spark](./spark.md#configuration) script.

```scala
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.access.key", "AKIAIOSFODNN7EXAMPLE")
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.secret.key", "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY")
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.endpoint", "https://s3.lakefs.example.com")
```

> Once set, you can now interact with Delta tables using the same commands you would use without lakeFS, except for altering the path prefix to include the lakeFS repo and branch names:

Reviewer: ```suggestion
Once set, you can now interact with Delta tables using regular Spark path URIs. Make sure you include the lakeFS repository and branch name:
```

```scala
data.write.format("delta").save("s3a://<repo-name>/<branch-name>/path/to/delta-table")
```

Note: If using the Databricks Analytics Platform, see the [integration guide](./databricks.md#configuration) for configuring a Databricks cluster to use lakeFS.

## Limitations

> The [Delta log](https://databricks.com/blog/2019/08/21/diving-into-delta-lake-unpacking-the-transaction-log.html) is an auto-generated text file used to keep track of transactions on a Delta table sequentially. Writing to one Delta table from multiple lakeFS branches is possible, but note that it will result in conflicts if later attempting to merge one branches into the other.

Reviewer: ```suggestion
The [Delta log](https://databricks.com/blog/2019/08/21/diving-into-delta-lake-unpacking-the-transaction-log.html) is an auto-generated sequence of text files used to keep track of transactions on a Delta table sequentially. Writing to one Delta table from multiple lakeFS branches is possible, but note that it will result in conflicts if later attempting to merge one branch into the other. For that reason, production workflows should ideally write to a single lakeFS branch that could then be safely merged into `main`. 
```

## Read more

> [Blog post](https://lakefs.io/guarantee-consistency-in-your-delta-lake-tables-with-lakefs/) that shows how to 

Reviewer: ```suggestion
See [this post on the lakeFS blog](https://lakefs.io/guarantee-consistency-in-your-delta-lake-tables-with-lakefs/) that shows how to 
```

---
layout: default

> title: Delta

Reviewer: ```suggestion
title: Delta Lake
```

> description: This section explains how to interact with your lakeFS when using delta format

Reviewer: ```suggestion
description: This section explains how to use Delta Lake with lakeFS
```

parent: Integrations
nav_order: 62
has_children: false
---

# Using lakeFS with Delta Lake
{: .no_toc }

> [Delta Lake](https://delta.io/) is an open source project that enables building a Lakehouse architecture on top of data lakes.  

Reviewer: ```suggestion
[Delta Lake](https://delta.io/) Delta lake is an open-source project that enables building a Lakehouse architecture on top of existing storage systems such as S3, ADLS, GCS, and HDFS.
```

## Table of contents
{: .no_toc .text-delta }

1. TOC
{:toc}
   
## Use-cases
Using lakeFS with Delta tables allows supercharging Delta Lake to enable much better data safety guarantees, while simplifying operations:

1. ACID operations can now span across many Delta tables.
1. [CI/CD hooks](../setup/hooks.md) can be used to validate data quality and even ensure referential integrity.
1. Zero-copy [development environment](../usecases/data-devenv.md): Experiment with ETL code, algorithm changes and schema changes in isolation, without copying data!

## Configuration

> `Delta` is a data source format and files are being written as Parquet files by the s3a driver.

Reviewer: This sentence is unclear and I'm not sure it belongs here?

We need to set the proper credentials and endpoint in the S3 Hadoop configuration,
like we do with any [Spark](./spark.md#configuration) script.

```scala
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.access.key", "AKIAIOSFODNN7EXAMPLE")
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.secret.key", "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY")
 sc.hadoopConfiguration.set("spark.hadoop.fs.s3a.bucket.<repo-name>.endpoint", "https://s3.lakefs.example.com")
```

> You can continue to read/write Delta tables the same way you did without lakeFS,

Reviewer: This assumes I used to read/write Delta tables before I had lakeFS which going forward might not be true..

while using lakeFS repos and branchs as the path prefixes:

```scala

> data.write.format("delta").save("s3a://<repo-name>/<branch-name>/tmp/delta-table")

Reviewer: `/tmp` is a bit confusing here

```

## Limitations
Delta logs are sequential. Writing to the same delta log from multiple lakeFS branches is possible,
But would result with conflicts when trying to merge those branches.

> It's best to avoid writing to the same Delta log from different branches.

Reviewer: Why avoid? For a dev env, this is absolutely OK. The caveat here is that I won't be able to merge.
